File: ads/src/main/java/com/example/ads/newStrategy/GoogleAppOpen.java
package com.example.ads.newStrategy;


import android.content.Context;

import androidx.annotation.NonNull;

import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.LoadAdError;
import com.google.android.gms.ads.appopen.AppOpenAd;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Stack;

import timber.log.Timber;

public class GoogleAppOpen {

    final private int totalLevels = 4;
    private ArrayList<ArrayList<Object>> adUnits;

    private final String adUnitId = "ca-app-pub-3940256099942544/3419835294";

    public GoogleAppOpen(Context context) {
        instantiateList();
        loadAppopenStart(context);
    }


    private void instantiateList() {
        adUnits = new ArrayList<>();

        adUnits.add(0, new ArrayList<Object>(Arrays.asList(adUnitId, new Stack<AppOpenAd>())));
        adUnits.add(1, new ArrayList<Object>(Arrays.asList(adUnitId, new Stack<AppOpenAd>())));
        adUnits.add(2, new ArrayList<Object>(Arrays.asList(adUnitId, new Stack<AppOpenAd>())));
        adUnits.add(3, new ArrayList<Object>(Arrays.asList(adUnitId, new Stack<AppOpenAd>())));
        adUnits.add(4, new ArrayList<Object>(Arrays.asList(adUnitId, new Stack<AppOpenAd>())));
    }

    public void loadAppopenStart(Context context) {
        AppOpenAdLoad(context, totalLevels);
    }


    public AppOpenAd getAd(Context activity) {
        for (int i = totalLevels; i >= 0; i--) {
            ArrayList<Object> list = adUnits.get(i);
            String adunitid = (String) list.get(0);
            Stack<AppOpenAd> stack = (Stack<AppOpenAd>) list.get(1);
            AppOpenLoadSpecific(activity, adunitid, stack);
            if (stack != null && !stack.isEmpty()) {
                return stack.pop();
            }
        }

        return null;
    }

    public void AppOpenAdLoad(Context activity, int level) {
        if (level < 0) {
            return;
        }

        if (adUnits.size() < level) {
            return;
        }

        ArrayList<Object> list = adUnits.get(level);
        String adunitid = (String) list.get(0);
        Stack<AppOpenAd> stack = (Stack<AppOpenAd>) list.get(1);

        AdRequest request = new AdRequest.Builder().build();
        AppOpenAd.load(activity, adunitid, request, AppOpenAd.APP_OPEN_AD_ORIENTATION_PORTRAIT, new AppOpenAd.AppOpenAdLoadCallback() {
            @Override
            public void onAdLoaded(@NonNull AppOpenAd ad) {
                stack.push(ad);
            }

            @Override
            public void onAdFailedToLoad(@NonNull LoadAdError loadAdError) {
                AppOpenAdLoad(activity, level - 1);
            }
        });
    }

    public void AppOpenLoadSpecific(Context activity, String adUnitId, Stack<AppOpenAd> stack) {
        AdRequest request = new AdRequest.Builder().build();

        AppOpenAd.load(activity, adUnitId, request, AppOpenAd.APP_OPEN_AD_ORIENTATION_PORTRAIT, new AppOpenAd.AppOpenAdLoadCallback() {
            @Override
            public void onAdLoaded(@NonNull AppOpenAd ad) {
                stack.push(ad);
            }

            @Override
            public void onAdFailedToLoad(@NonNull LoadAdError loadAdError) {
            }
        });
    }

}

File: ads/src/main/java/com/example/ads/newStrategy/GoogleAppOpen.kt
package com.example.ads.newStrategy

import android.content.Context
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.appopen.AppOpenAd
import com.google.android.gms.ads.appopen.AppOpenAd.AppOpenAdLoadCallback
import java.util.Stack

class GoogleAppOpen(context: Context?) {
    private val totalLevels = 4
    private var adUnits: ArrayList<ArrayList<Any>>? = null
    private val appOpen5 = "ca-app-pub-9507635869843997/7348324907"
    private val appOpen4 = "ca-app-pub-9507635869843997/2503192334"
    private val appOpenHigh = "ca-app-pub-9507635869843997/6418386614"
    private val appOpenMed = "ca-app-pub-9507635869843997/6226814923"
    private val appOpenAll = "ca-app-pub-9507635869843997/2287569917"

    init {
        instantiateList()
        loadAppopenStart(context)
    }

    private fun instantiateList() {
        adUnits = ArrayList()
        adUnits!!.add(0, ArrayList(listOf(appOpen5, Stack<AppOpenAd>())))
        adUnits!!.add(1, ArrayList(listOf(appOpen4, Stack<AppOpenAd>())))
        adUnits!!.add(2, ArrayList(listOf(appOpenHigh, Stack<AppOpenAd>())))
        adUnits!!.add(3, ArrayList(listOf(appOpenMed, Stack<AppOpenAd>())))
        adUnits!!.add(4, ArrayList(listOf(appOpenAll, Stack<AppOpenAd>())))
    }

    fun loadAppopenStart(context: Context?) {
        AppOpenAdLoad(context, totalLevels)
    }

    fun getAd(activity: Context?): AppOpenAd? {
        for (i in totalLevels downTo 0) {
            val list = adUnits!![i]
            val adunitid = list[0] as String
            val stack = list[1] as Stack<AppOpenAd>
            AppOpenLoadSpecific(activity, adunitid, stack)
            if (stack != null && !stack.isEmpty()) {
                return stack.pop()
            }
        }
        return null
    }

    fun AppOpenAdLoad(activity: Context?, level: Int) {
        if (level < 0) {
            return
        }
        if (adUnits!!.size < level) {
            return
        }
        val list = adUnits!![level]
        val adunitid = list[0] as String
        val stack = list[1] as Stack<AppOpenAd>
        val request = AdRequest.Builder().build()
        AppOpenAd.load(
            activity,
            adunitid,
            request,
            AppOpenAd.APP_OPEN_AD_ORIENTATION_PORTRAIT,
            object : AppOpenAdLoadCallback() {
                override fun onAdLoaded(ad: AppOpenAd) {
                    stack.push(ad)
                }

                override fun onAdFailedToLoad(loadAdError: LoadAdError) {
                    AppOpenAdLoad(activity, level - 1)
                }
            })
    }

    fun AppOpenLoadSpecific(activity: Context?, adUnitId: String?, stack: Stack<AppOpenAd>?) {
        val request = AdRequest.Builder().build()
        AppOpenAd.load(
            activity,
            adUnitId,
            request,
            AppOpenAd.APP_OPEN_AD_ORIENTATION_PORTRAIT,
            object : AppOpenAdLoadCallback() {
                override fun onAdLoaded(ad: AppOpenAd) {
                    stack!!.push(ad)
                }

                override fun onAdFailedToLoad(loadAdError: LoadAdError) {}
            })
    }
}